### DISCUSSION TITLE
Poor performance of CustomAFM model using deepctr_torch.models AFM on test data

### DISCUSSION THREAD

**Discussion by linkedlist771** [autogluon/autogluon#3122] [Q&A]
I have implemented a custom AutoGluon model using the AFM model from the `deepctr_torch.models` library. The model seems to train correctly, with the training loss gradually decreasing during the training process. However, I noticed that the validation score and test score are both very low, indicating poor performance on test data.

Below is the code snippet of my implementation:

<details>
  <summary>Click to expand the code snippet</summary>
  
  ```python

batch_size = 256
epochs = 1
verbose = 2
validation_split = 0

class CustomAFMModel(AFM):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)

    def predict(self, _test_input, batch_size=512):
        if isinstance(_test_input, pd.DataFrame):
            _test_input = CustomAFM.custom_instance.get_input(_test_input)
        predictions = super().predict(_test_input, batch_size).reshape(-1)
        return predictions

class CustomAFM(AbstractModel):
    custom_instance = None
    def __init__(self, **kwargs):
        super().__init__(**kwargs)
        self.model_cls = CustomAFMModel
        self.label = None
        self.linear_feature_columns = None
        self.dnn_feature_columns = None
        self.feature_names = None
        self.df = pd.read_excel(r"data.xlsx")
        CustomAFM.custom_instance = self

    def get_input(self, X: pd.DataFrame):

        print(self.df.shape)
        label = "10 mA cm‐2 "
        if label in X.columns:
            X = X.drop(columns=[label])
        sparse_features = X.columns
        X_index = X.index
        df = self.df.copy()
        df = df[sparse_features]
        # 1.Label Encoding for sparse features,and do simple Transformation for dense features
        for feat in sparse_features:
            lbe = LabelEncoder()
            df[feat] = lbe.fit_transform(df[feat])
        X_index_df_in_df = df.iloc[X_index]
        # 2.count #unique features for each sparse field
        fixlen_feature_columns = [SparseFeat(feat, df[feat].nunique())
                                  for feat in sparse_features]
        self.linear_feature_columns = fixlen_feature_columns
        print(f"feature_size: {len(self.linear_feature_columns)}")
        self.dnn_feature_columns = self.linear_feature_columns
        self.feature_names = get_feature_names(self.linear_feature_columns + self.dnn_feature_columns)
        df = df.iloc[X_index]
        train_model_input = {name: df[name].values for name in self.feature_names}
        return train_model_input

    def _fit(self, X: pd.DataFrame, y: pd.Series, **kwargs):
        print('Entering the `_fit` method')
        train_model_input = self.get_input(X)
        device = 'cpu'
        use_cuda = True
        if use_cuda and torch.cuda.is_available():
            print('cuda ready...')
            device = 'cuda:0'

        model = self.model_cls(linear_feature_columns=self.linear_feature_columns,
                                 dnn_feature_columns=self.dnn_feature_columns, task='regression', device=device)
        print("modle init done")
        model.compile("adam", "mse", metrics=['mse'], )
        self.model = model
        history = model.fit(train_model_input, y.values,
                            batch_size=batch_size,
                            epochs=epochs,
                            verbose=verbose,
                            validation_split=validation_split
                            )
        print('Exiting the `_fit` method')

    def _set_default_params(self):
        for param, val in self.params.items():
            self._set_default_param_value(param, val)

    # The `_get_default_auxiliary_params` method defines various model-agnostic parameters such as maximum memory usage and valid input column dtypes.
    # For most users who build custom models, they will only need to specify the valid/invalid dtypes to the model here.
    def _get_default_auxiliary_params(self) -> dict:
        default_auxiliary_params = super()._get_default_auxiliary_params()
        extra_auxiliary_params = dict(
            # the total set of raw dtypes are: ['int', 'float', 'category', 'object', 'datetime']
            # object feature dtypes include raw text and image paths, which should only be handled by specialized models
            # datetime raw dtypes are generally converted to int in upstream pre-processing,
            # so models generally shouldn't need to explicitly support datetime dtypes.
            valid_raw_types=['int', 'float', 'category'],
            # Other options include `valid_special_types`, `ignored_type_group_raw`, and `ignored_type_group_special`.
            # Refer to AbstractModel for more details on available options.
        )
        default_auxiliary_params.update(extra_auxiliary_params)
        return default_auxiliary_params

    def predict(self, x: pd.DataFrame):
        print("Entering the `predict` method")
        print(x.shape)
        print(x)
        return self.model.predict(x)
```
</details>

## Questions / Need help with

I would like to ask for guidance or suggestions on how to improve the performance of my custom model on test data. Specifically, I am interested in:

- Understanding if there are any issues in the implementation that might be causing the poor performance on test data.
- Suggestions on possible hyperparameter tuning or model architecture adjustments that could potentially improve the model's performance.
- Any other general tips or best practices for using deepctr_torch models in custom AutoGluon models.
